### DISCUSSION TITLE
Multi-class  classification

### DISCUSSION THREAD

**Discussion by morgand-sql** [mrdbourke/pytorch-deep-learning#857] [Q&A]
video lesson 87 .
topic :  Creating a training and testing loop for a multi-class PyTorch model
problem : I am getting results that are so weird . I need to understand why

Epoch: 0 | Loss: 1.04324, Acc: 65.50% | Test Loss: 0.57861, Test Acc: 95.50%
Epoch: 10 | Loss: 0.05260, Acc: 98.75% | Test Loss: 0.00351, Test Acc: 100.00%
Epoch: 20 | Loss: 0.15444, Acc: 99.00% | Test Loss: 0.00025, Test Acc: 100.00%
Epoch: 30 | Loss: 0.16679, Acc: 99.25% | Test Loss: 0.05451, Test Acc: 99.50%
Epoch: 40 | Loss: 24442644.00000, Acc: 26.12% | Test Loss: 4618876616704.00000, Test Acc: 31.50%
Epoch: 50 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 60 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 70 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 80 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 90 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%

**Comment by LuluW8071** [comment]
You need to post your code snippet as well can't determine anything from your output alone.

**Reply by morgand-sql** [reply]
# fit the model 
torch.manual_seed(42)

## set the number of epochs
epochs=100

## put data into the target device

X_blob_train,y_blob_train= X_blob_train.to(device),y_blob_train.to(device)
X_blob_test,y_blob_test=X_blob_test.to(device),y_blob_test.to(device)


for epoch in range(epochs):
  ### training
  model_4.train()

  #1. forawrd pass
  y_logits=model_4.linear_layer_stack(X_blob_train)
  y_pred=torch.softmax(y_logits,dim=1,).argmax(dim=1)

  #2.0 calculate the loss

  loss=loss_fn(y_logits,y_blob_train)
  acc=accuracy_fn(y_true=y_blob_train,
                  y_pred=y_pred)
  
  #3. loss backward
  loss.backward()

  

  #5.0 optimizer step
  optimizer.step()

  ### testing 
  model_4.eval()

  with torch.inference_mode():
    ### 1.0 forward pass
    test_logits=model_4.linear_layer_stack(X_blob_test)
    test_pred=torch.softmax(test_logits,dim=1).argmax(dim=1)

    ## 2. calculate the loss
    test_loss=loss_fn(test_logits,y_blob_test)
    test_acc=accuracy_fn(y_true=y_blob_test,
                         y_pred=test_pred)
    
    ## print out whats happening 
  if epoch % 10 ==0:
     print(f"Epoch: {epoch} | Loss: {loss:.5f}, Acc: {acc:.2f}% | Test Loss: {test_loss:.5f}, Test Acc: {test_acc:.2f}%")

**Reply by morgand-sql** [reply]
this is my code @LuluW8071

**Reply by LuluW8071** [reply]
```python
model_4.train()

# 1. forward pass
y_logits=model_4(X_blob_train)
y_pred=torch.softmax(y_logits,dim=1,).argmax(dim=1)

# 2. calculate the loss
loss=loss_fn(y_logits,y_blob_train)
acc=accuracy_fn(y_true=y_blob_train,
            y_pred=y_pred)

# 3. Optimizer Zero grad
optimizer.zero_grad()

# 3. loss backward
loss.backward()

# 5. optimizer step
optimizer.step()

```

You forgot to add `optimizer.zero_grad()`

Also why are you forward passing `y_logits=model_4.linear_layer_stack(X_blob_train)`
It should be `y_logits=model_4(X_blob_train)`
Your testing code also should be corrected `test_logits=model_4(X_blob_test)`

**Reply by morgand-sql** [reply]
AttributeError                            Traceback (most recent call last)
[<ipython-input-439-09c3e3e35591>](https://localhost:8080/#) in <cell line: 11>()
     14 
     15     # 1. Forward pass
---> 16     y_logits = model_4(X_blob_train) # model outputs raw logits
     17     y_pred = torch.softmax(y_logits, dim=1).argmax(dim=1) # go from logits -> prediction probabilities -> prediction labels
     18     # print(y_logits)

3 frames
[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in __getattr__(self, name)
   1693             if name in modules:
   1694                 return modules[name]
-> 1695         raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")
   1696 
   1697     def __setattr__(self, name: str, value: Union[Tensor, 'Module']) -> None:

AttributeError: 'BlobModel' object has no attribute 'layer_stack 



this the error I get when I don't implement layer_stack

**Reply by morgand-sql** [reply]
Epoch: 0 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 10 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 20 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 30 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 40 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 50 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 60 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 70 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 80 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%
Epoch: 90 | Loss: nan, Acc: 25.12% | Test Loss: nan, Test Acc: 24.50%


still getting this with those chamges you highlighted

**Reply by LuluW8071** [reply]
I guess you implemented your model_4 class in different way not like in the video. If so that explains why the error came. Do same as before just add `optimizer.zero_grad()`

Can you upload ur notebook here or in your public repo.

**Reply by morgand-sql** [reply]
I am having problems uploading but it should be here in few minutes

**Reply by morgand-sql** [reply]
[NEURAL NETWORK CLASSIFICATION.zip](https://githu@LuluW8071 
.com/mrdbourke/pytorch-deep-learning/files/14546144/NEURAL.NETWORK.CLASSIFICATION.zip)
@LuluW8071 this is my notebook

**Reply by LuluW8071** [reply]
```python
class BlobModel(nn.Module):
  def __init__(self, input_features,output_features, hidden_units=8):
        super().__init__()
        self.linear_layer_stack = nn.Sequential(
            nn.Linear(in_features=input_features, out_features=hidden_units),
            nn.Linear(in_features=hidden_units, out_features=hidden_units),
            nn.Linear(in_features=hidden_units, out_features=output_features))

  def forward(self,x):
        # You wrote only self.layer_stack instead of self.linear_layer_stack
        return self.linear_layer_stack(x)  
```
Now change forward passing `y_logits=model_4.linear_layer_stack(X_blob_train)` to `y_logits=model_4(X_blob_train)`

and perform similar actions to testing code `test_logits=model_4(X_blob_test)`

Do same for above training code for forward passing of example logits

If your accuracy is low define another model for non-linearity and training and testing loop is same

```python
class BlobModel(nn.Module):
  def __init__(self, input_features,output_features, hidden_units=8):
        super().__init__()
        self.linear_layer_stack = nn.Sequential(
            nn.Linear(in_features=input_features, out_features=hidden_units),
            nn.ReLU(),
            nn.Linear(in_features=hidden_units, out_features=hidden_units),
            nn.ReLU(),
            nn.Linear(in_features=hidden_units, out_features=output_features))

  def forward(self,x):
        # You wrote only self.layer_stack instead of self.linear_layer_stack
        return self.linear_layer_stack(x) 
  ```
  
  Also do note that sometimes passing `X` instead of `x` can cause conflict in training may be incorrect results while forward passing in model class

**Reply by LuluW8071** [reply]
## With no `ReLU()`
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/107304848/927f7f15-df95-4a1f-8c82-322e1d7aa77b)

## With `ReLU()`
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/107304848/e88ce333-d6e6-4e6d-b241-421b80641ac5)

**Reply by morgand-sql** [reply]
@LuluW8071 sorry for taking time to get back to you . I have made those changes and it worked . I really appreciate your effort , that was so swift.
